docs/vfio-user.rst

```diff
@@ -1131,6 +1144,168 @@ client must write data on the same order and transction size as read.
+| Flags        | Reply bit set in reply |
++--------------+------------------------+
+| Error        | 0/errno                |
++--------------+------------------------+
+| Region info  | Region IO FD info      |
++--------------+------------------------+
+
+Clients can access regions via VFIO_USER_REGION_READ/WRITE or, if provided, by
+mmap()ing a file descriptor provided by the server.
+
+VFIO_USER_DEVICE_GET_REGION_IO_FDS provides an alternative access mechanism via
+file descriptors. This is an optional feature intended for performance
+improvements where an underlying sub-system (such as KVM) supports communication
+across such file descriptors to the vfio-user server, without needing to
+round-trip through the client.
+
+The server returns an array describing sub-regions of the given region along
+with the server specifies a set of sub-regions and the requested file descriptor
+notification mechanism to use for that sub-region.  Each sub-region in the
+response message may choose to use a different method, as defined below.  The
+two mechanisms supported in this specification are ioeventfds and ioregionfds.
+
+A client should then hook up the returned file descriptors with the notification
+method requested.
+
+Region IO FD info format
+^^^^^^^^^^^^^^^^^^^^^^^^
+
++------------+--------+------+
+| Name       | Offset | Size |
++============+========+======+
+| argsz      | 16     | 4    |
++------------+--------+------+
+| flags      | 20     | 4    |
++------------+--------+------+
+| index      | 24     | 4    |
++------------+--------+------+
+| count      | 28     | 4    |
++------------+--------+------+
+| subregions | 32     | ...  |
++------------+--------+------+
+
+* *argsz* is the size of the region IO FD info structure plus the
+  total size of the subregion array. Thus, each array entry "i" is at offset
+  i * ((argsz - 32) / count). Note that currently this is 40 bytes for both IO
+  FD types, but this is not to be relied on.
+* *flags* must be zero
+* *index* is the index of memory region being queried
+* *count* is the number of sub-regions in the array
+* *subregions* is the array of Sub-Region IO FD info structures
+
+The client must set ``flags`` to zero and specify the region being queried in
+the ``index``.
+
+The client sets the ``argsz`` field to indicate the maximum size of the response
+that the server can send, which must be at least the size of the response header
+plus space for the subregion array. If the full response size exceeds ``argsz``,
+then the server must respond only with the response header and the Region IO FD
+info structure, setting in ``argsz`` the buffer size required to store the full
+response. In this case, no file descriptors are passed back.  The client then
+retries the operation with a larger receive buffer.
+
+The reply message will additionally include at least one file descriptor in the
+ancillary data. Note that more than one subregion may share the same file
+descriptor.
```

Comment:
I think it's worth adding that in Linux sharing the same file descriptor multiple times results in seemingly different file descriptors in the message.

Reply:
That's not true in this context, server can pass back just one fd, and have all sub-regions use the same fd_index value.